# Detailed Mutation Analysis Tutorial Using PyEED

This notebook demonstrates a comprehensive workflow for analyzing mutations between protein sequences using PyEED (Python Enzyme Engineering Database). We'll walk through each step in detail.

## Cell 1: Required Imports

In [7]:
import logging

from pyeed import Pyeed
from pyeed.analysis.mutation_detection import MutationDetection
from pyeed.analysis.standard_numbering import StandardNumberingTool

- `Pyeed`: Main class for interacting with the PyEED database
- `MutationDetection`: Class for identifying differences between protein sequences
- `StandardNumberingTool`: Ensures consistent position numbering across different protein sequences
- `logging`: For tracking execution progress and debugging

## Cell 2: Logging Configuration

In [8]:
logging.basicConfig(level=logging.ERROR, format='%(asctime)s - %(levelname)s - %(message)s')
LOGGER = logging.getLogger(__name__)

Sets up logging to:
- Show only ERROR level messages and above
- Include timestamp, log level, and message in the output
- Create a logger instance specific to this notebook

## Cell 3: Database Setup and Connection

In [9]:
uri = "bolt://129.69.129.130:7687"
user = "neo4j"
password = "12345678"

eedb = Pyeed(uri, user=user, password=password)

eedb.db.wipe_database("2025-01-19")
eedb.db.remove_db_constraints(user=user, password=password)

Pyeed Graph Object Mapping constraints not defined. Use _install_labels() to set up model constraints.
📡 Connected to database.
All data has been wiped from the database.
Connecting to bolt://neo4j:12345678@127.0.0.1:7687
Dropping constraints...

Dropping indexes...

All constraints and indexes have been removed from the database.


1. Establishes connection parameters to a local Neo4j database
2. Creates a PyEED instance with these credentials
3. Wipes existing database data (with date "2025-01-19")
4. Removes all database constraints for a fresh start

This ensures we're working with a clean database state.

## Cell 4: Sequence Retrieval

In [10]:
ids = ["KJO56189.1", "KLP91446.1"]

eedb.fetch_from_primary_db(ids, db='ncbi_protein')

2025-01-19 13:31:21.278 | INFO     | pyeed.main:fetch_from_primary_db:85 - Found 0 sequences in the database.
2025-01-19 13:31:21.279 | INFO     | pyeed.main:fetch_from_primary_db:87 - Fetching 2 sequences from ncbi_protein.
2025-01-19 13:31:21.317 | INFO     | pyeed.adapter.primary_db_adapter:make_request:157 - Making requests with ids list: ['KJO56189.1,KLP91446.1']
2025-01-19 13:31:21.317 | DEBUG    | pyeed.adapter.primary_db_adapter:make_request:161 - Sending 1 requests in batches of 10
2025-01-19 13:31:21.319 | DEBUG    | pyeed.adapter.primary_db_adapter:send_request:141 - Sending request to https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi with parameters: {'retmode': 'text', 'rettype': 'genbank', 'db': 'protein', 'id': 'KJO56189.1,KLP91446.1'}
2025-01-19 13:31:21.944 | DEBUG    | pyeed.adapter.primary_db_adapter:sanitize_response:196 - Response content: <class 'bytes'>
2025-01-19 13:31:21.946 | DEBUG    | pyeed.adapter.primary_db_adapter:make_request:175 - Received respo

1. Defines two protein sequence IDs to analyze
2. Fetches these sequences from NCBI's protein database
3. Both sequences are beta-lactamase proteins:
   - KJO56189.1: beta-lactamase TEM
   - KLP91446.1: class A beta-lactamase
4. The sequences are automatically parsed and stored in the Neo4j database
5. Additional metadata like organism information and CDS (Coding Sequence) details are also stored

## Cell 5: Standard Numbering Application

In [11]:
sn = StandardNumberingTool(name="test_standard_numbering")
sn.apply_standard_numbering(base_sequence_id='KJO56189.1', db=eedb.db, list_of_seq_ids=ids)

2025-01-19 13:31:22.058 | INFO     | pyeed.analysis.standard_numbering:apply_standard_numbering:342 - Using 2 sequences for standard numbering
2025-01-19 13:31:22.204 | INFO     | pyeed.analysis.standard_numbering:run_numbering_algorithm:72 - Running numbering algorithm for base sequence KJO56189.1


CLUSTAL O(1.2.4) multiple sequence alignment


KJO56189.1      MSIQHFRVALIPFFAAFCLPVFAHPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRP
KLP91446.1      MSIQHFRVALIPFFAAFCLPVFAHPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRP
                ************************************************************

KJO56189.1      EERFPMMSTFKVLLCGAVLSRVDAGQEQLGRRIHYSQNDLVEYSPVTEKHLTDGMTVREL
KLP91446.1      EERFPMMSTFKVLLCGAVLSRVDAGQEQLGRRIHYSQNDLVKYSPVTEKHLTDGMTVREL
                *****************************************:******************

KJO56189.1      CSAAITMSDNTAANLLLTTIGGPKELTAFLHNMGDHVTRLDSWEPELNEAIPNDERDTTM
KLP91446.1      CSAAITMSDNTAANLLLTTIGGPKELTAFLHNMGDHVTRLDRWEPELNEAIPNDERDTTM
                ***************************************** ******************

KJO56189.1      PAAMATTLRKLLTGELLTLASRQQLIDWMEADKVAGPLLRSALPAGWFIADKSGAGERGS
KLP91446.1      PAAMATTLRKLLTGELLTLASRQQLIDWMEADKVAGPLLRSALPAGWFIADKSGASERGS
                *******************************************************.****

KJO56189.1      RGIIAALGP

1. Creates a new StandardNumberingTool instance named "test_standard_numbering"
2. Uses KJO56189.1 as the reference sequence for numbering
3. Performs multiple sequence alignment (MSA) using CLUSTAL
4. The alignment output shows:
   - Asterisks (*) indicate identical residues
   - Colons (:) indicate conserved substitutions
   - Periods (.) indicate semi-conserved substitutions
5. This step is crucial for ensuring mutations are correctly identified relative to consistent positions

## Cell 6: Mutation Detection

In [12]:
md = MutationDetection()

seq1 = "KJO56189.1"
seq2 = "KLP91446.1"
name_of_standard_numbering_tool = "test_standard_numbering"

mutations = md.get_mutations_between_sequences(
    seq1, seq2, eedb.db, name_of_standard_numbering_tool
)

2025-01-19 13:31:22.519 | DEBUG    | pyeed.analysis.mutation_detection:save_mutations_to_db:135 - Saved 3 mutations to database


1. Creates a MutationDetection instance
2. Compares the two sequences using the standard numbering scheme
3. Identifies all positions where amino acids differ
4. Automatically saves the mutations to the database
5. Returns a dictionary containing mutation information

## Cell 7: Results Analysis

In [13]:
print(mutations)

{'from_positions': [102, 162, 236], 'to_positions': [102, 162, 236], 'from_monomers': ['E', 'S', 'G'], 'to_monomers': ['K', 'R', 'S']}


Outputs a detailed mutation map showing:
- `from_positions`: [102, 162, 236] - Where mutations occur in the sequence
- `to_positions`: [102, 162, 236] - Corresponding positions in the second sequence
- `from_monomers`: ['E', 'S', 'G'] - Original amino acids
- `to_monomers`: ['K', 'R', 'S'] - Mutated amino acids

This means we found three mutations:
1. Position 102: Glutamic acid (E) → Lysine (K)
2. Position 162: Serine (S) → Arginine (R)
3. Position 236: Glycine (G) → Serine (S)

These mutations could be significant for:
- Understanding protein evolution
- Analyzing functional differences
- Planning protein engineering experiments
- Studying antibiotic resistance mechanisms (since these are beta-lactamase proteins)

## Technical Notes:
- The database operations are performed using Neo4j's Bolt protocol
- Sequence data is retrieved using NCBI's E-utilities
- Multiple sequence alignment is performed using CLUSTAL
- All operations are tracked with detailed logging
- The system automatically handles protein metadata and relationships